engine/src/main/scala/cromwell/engine/workflow/WorkflowStoreActor.scala

```diff
@@ -0,0 +1,145 @@
+import cromwell.services.{MetadataEvent, MetadataKey, MetadataValue, ServiceRegistryClient}
+
+import scala.language.postfixOps
+import scala.collection._
+
+object WorkflowStore {
+  sealed trait WorkflowState {
+    def isStartable: Boolean = {
+      this match {
+        case x: StartableState => true
+        case _ => false
+      }
+    }
+  }
+
+  case object Running extends WorkflowState
+
+  sealed trait StartableState extends WorkflowState
+  case object Submitted extends StartableState
+  case object Restartable extends StartableState
+
+  final case class SubmittedWorkflow(sources: WorkflowSourceFiles, state: WorkflowState) {
+    def toWorkflowToStart(id: WorkflowId): WorkflowToStart = {
+      state match {
+        case r: StartableState => WorkflowToStart(id, sources, r)
+        case _ => throw new IllegalArgumentException("This workflow is not currently in a startable state")
+      }
+    }
+  }
+
+  final case class WorkflowToStart(id: WorkflowId, sources: WorkflowSourceFiles, state: StartableState)
+}
+
+/**
+  * Stores any submitted workflows which have not completed.
+  *
+  * This is not thread-safe, although it should not be being used in a multithreaded fashion. If that
+  * changes, revisit.
+  */
+trait WorkflowStore {
+  import WorkflowStore._
+
+  /*
+    The LinkedHashMap preserves insertion order which means that any requests for runnable workflows will
+    be satisfied by taking the oldest available.
+   */
+  private var workflowStore = new mutable.LinkedHashMap[WorkflowId, SubmittedWorkflow]
+
+  /**
+    * Adds a WorkflowSourceFiles to the store and returns a WorkflowId for tracking purposes
+    */
+  def add(source: WorkflowSourceFiles): WorkflowId = {
+    val workflowId = WorkflowId(UUID.randomUUID())
+    workflowStore += (workflowId -> SubmittedWorkflow(source, Submitted))
+    workflowId
+  }
+
+  /**
+    * Provides a bulk add functionality. Adds all of the provided WorkflowSourceFiles at once, returning a List
+    * of UUIDs matching the order of the supplied source files
+    */
+  def add(sources: Seq[WorkflowSourceFiles]): immutable.List[WorkflowId] = {
+    val submittedWorkflows = sources.toList map { WorkflowId(UUID.randomUUID()) -> SubmittedWorkflow(_, Submitted) }
+    workflowStore ++= submittedWorkflows
+    submittedWorkflows map { _._1 }
+  }
+
+  /**
+    * Retrieves up to n workflows which have not already been pulled into the engine and sets their pickedUp
+    * flag to true
+    */
+  def fetchRunnableWorkflows(n: Int): immutable.List[WorkflowToStart] = {
+    val startableWorkflows = workflowStore filter { case (k, v) => v.state.isStartable } take n
+    workflowStore ++= startableWorkflows map { case (k, v) => k -> SubmittedWorkflow(v.sources, Running) }
+    // Because we explicitly filtered out for startable states this is safe
+    startableWorkflows map { case (k, v) => v.toWorkflowToStart(k) } toList
+  }
+
+  // FIXME: Do we care if it doesn't exist? If so to what extent
```

Comment:
I was logging it for a while but took it out because both a) there are already spots in the WMA where it also silently soldiers on in cases like this and b) It seems like the sort of thing that no one would ever notice anyways. 

It's easy enough to add a log in there, it's of the This Shouldn't Happen category but not enough to blow the system up IMO